### PR TITLE
Add RaceID3 StemID2 (FateID) Package

### PR DESCRIPTION
**Pull request by mtekman**
FOR CONTRIBUTOR:
* [X] - I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [X] - License permits unrestricted use (educational + commercial)
* [ ] - This PR adds a new tool or tool collection
* [X] - This PR updates an existing tool or tool collection
* [ ] - This PR does something else (explain below)

FOR REVIEWER:
* [ ] .shed.yml file ok
    - [ ] Toolshed user `iuc` has access to associated toolshed repo(s)
* [ ] Indentation is correct (4 spaces)
* [ ] Tool version/build ok
* [ ] `<command/>`
  - [ ] Text parameters, input and output files `'single quoted'`
  - [ ] Use of `<![CDATA[ ... ]]>` tags
  - [ ] Parameters of type `text` or having `optional="true"` attribute are checked with `if str($param)` before being used
* [ ] Data parameters have a `format` attribute containing datatypes recognised by Galaxy
* [ ] Tests
  - [ ] Parameters are reasonably covered
  - [ ] Test files are appropriate
* [ ] Help
  - [ ] Valid restructuredText and uses `<![CDATA[ ... ]]>` tags
* [ ] Complies with other best practice in [Best Practices Doc](http://galaxy-iuc-standards.readthedocs.io/en/latest/best_practices/tool_xml.html)

****************************

INFO ABOUT PACKAGE

This tool is an update to the old RaceID package, which has had a major update resulting in the loss and gain of new features. As a result, a major re-write of the wrapper is required and the aim is for the tool(s) here to completely replace the existing, with the existing tool being retired.

 * The tool is split into 4 modules:
   1. RaceID Cluster - Performs general clustering
   1. RaceID Cluster Inspect - Allows for the inspection of the above cluster
   1. RaceID Trajectory - Performs pseudo-temporal lineage analysis upon the cluster
   1. RaceID Trajectory Inspect - Allows for the inspection of the above lineages.

 * Each module contains all the important parameters outlined in the vignette
 * Edit: All modules are now merged into one tool

GENERAL TODO LIST

 1. Cluster
    * [x] R script
    * [x] General XML
    * [x] Cheetah
    * [x] Tests
      * [x] Defaults
      * [x] Advanced1
      * [x] Advanced2

 1. Cluster Inspect
    * [x] R script
    * [x] General XML
    * [x] Cheetah
    * [x] Tests
      * [x] Defaults
      * [x] Advanced1
      * [x] Advanced2

 1. Trajectory
    * [x] R script
    * [x] General XML
    * [x] Cheetah
    * [x] Tests
      * [x] Defaults
      * [x] Advanced1
      * [x] Advanced2

 1. Trajectory Inspect
    * [x] R script
    * [x] General XML
    * [x] Cheetah
    * [x] Tests
      * [x] Defaults
      * [x] Advanced1
      * [x] Advanced2
      * [x] Advanced3

 1. Other
    * [x] Fix bug preventing Galaxy from loading r-raceid bioconda env
          - small bug with glibc and `later` package in arch linux
    * [x] Verify that tests don't crash
    * [x] Merge tools into one
    * [x] Description & Help
       * [x] Capitalise labels
       * [x] Better help text for all parameters and RDS meaning
       * [x] Better help text in section tags (e.g. FateID and StemID)
    * [x] Clean up Sections followed immediately by non default conditional
    * [ ] (Optional) Add LOOM support

### CODE REVIEW

**Comment by mvdbeek** [comment]
@mtekman you will need to rebase this PR onto the current master, we had accidentally added a large commit that I rebased out of the master branch now. Let me know if you need any help or if I should do this for you.

File: tools/raceid3/.shed.yml [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+remote_repository_url: https://github.com/galaxyproject/tools-iuc/tree/master/tools/raceid3
+long_description: |
+  RaceID is a method for cell type identification from single-cell RNA-seq data by unsupervised learning. An initial clustering is followed by an outlier identification based on a backgorund model of combined technical and biological variability in single-cell RNA-seq data obtained by quantification with unique molecular identifiers. StemID permits subsequent inference of a lineage tree based on clusters, i.e. cell types, identified by RaceID.
+owner: iuc
+type: unrestricted
+auto_tool_repositories:
+  name_template: "{{ tool_id }}"
+  description_template: "Wrapper for the RaceID pipeline tool: {{ tool_name }}"
+suite:
+  name: "suite_singlecell"
```
Comment (bgruening) [inline tools/raceid3/.shed.yml]: suite_raceid

File: tools/raceid3/.shed.yml [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  description: Single Cell utilities - Modules for the processing of scRNA count matrices
```
Comment (bgruening) [inline tools/raceid3/.shed.yml]: raceid?

Comment (bgruening) [reply tools/raceid3/.shed.yml]: singel cell is just to broad

File: tools/raceid3/macros.xml [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+        </conditional>
+    </macro>
+    <macro name="test_nondef" >
+        <conditional name="use" >
+            <param name="def" value="no" />
+            <yield />
+        </conditional>
+    </macro>
+    <macro name="section_test_nondef" >
+        <section name="@SECTION_NAME@" >
```
Comment (bgruening) [inline tools/raceid3/macros.xml]: Is this used? You are not using the macro inheritance as it seems?

File: tools/raceid3/macros.xml [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+}
+
+string2textvector = function(string){
+    #' Converts a string to a valid string vector
+    #'
+    #'   e.g. "red,yellow,white" evaluates to c("red","yellow","white")
+    return(unlist(strsplit(string,",")))
+}
+    </token>
+    <token name="@WRAPPER_VERSION@">0.1.0</token>
```
Comment (bgruening) [inline tools/raceid3/macros.xml]: 0.1.1 as the raceid pkg?

File: tools/raceid3/macros_inspectclusters.xml [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+#### (Optional) Specific set of genes
+
+#if str($ofc.gois.do_opt) == "yes":
+perform.genesofinterest = TRUE
+plotexp = formals(plotexpmap)
+plotmarkg = formals(plotmarkergenes)
+### Defaults
+## plotsym.use.typeremoveregex = "\\_\\d+"   -- selecting types
+## plotsym.use.typerremoveregex.subselect = "IV|V"
+## inspect.goi.use.genes = genes
```
Comment (bgruening) [inline tools/raceid3/macros_inspectclusters.xml]: should the comments be removed?

File: tools/raceid3/macros_inspecttrajectories.xml [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+                </section>
+            </conditional>
+            <output name="outdiffgenes" value="out_traject_inspect_allthree.tabular" />
+            <output name="outpdf" value="out_traject_inspect_allthree.pdf" compare="sim_size" delta="15" />
+        </test>
+    </macro>
+    <token name="@INSPECTTRAJECTORIES_CHEETAH@"><![CDATA[
+#set ofc = $tool
+
+in.rdat = readRDS("${ofc.inputrds}")
```
Comment (bgruening) [inline tools/raceid3/macros_inspecttrajectories.xml]: single-quotes?

File: tools/raceid3/macros_inspecttrajectories.xml [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+
+#if str($ofc.trjfid.basic.doit) == "yes":
+perform.fateID = TRUE
+trjfid.cellsfrom = formals(cellsfromtree)
+trjfid.filterset = formals(filterset)
+trjfid.getsom = formals(getsom)
+trjfid.procsom = formals(procsom)
+trjfid.plotheat = list()
+## trjfid.cellsfrom\$z = c(1,2,3,4) # a valid vector, not related to ijk cells
+## trjfid.filterset\$minexpr = 2
```
Comment (bgruening) [inline tools/raceid3/macros_inspecttrajectories.xml]: comments?

File: tools/raceid3/macros_trajectory.xml [hunk 8]
```diff
@@ -0,0 +1,181 @@
+<macros>
+    <macro name="trajectory_inputs" >
+        <param name="inputrds" type="data" format="rdata" label="Input RDS" />
```
Comment (bgruening) [inline tools/raceid3/macros_trajectory.xml]: write RDS out or explain what that is, where to get it in the help?

File: tools/raceid3/raceid3_main.xml [hunk 9]
```diff
@@ -0,0 +1,101 @@
+<tool id="raceid3_main" name="RaceID3" version="@WRAPPER_VERSION@" >
```
Comment (bgruening) [inline tools/raceid3/raceid3_main.xml]: just `raceid3` now?

File: tools/raceid3/raceid3_main.xml [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+        <import>macros_trajectory.xml</import>
+        <import>macros_inspecttrajectories.xml</import>
+    </macros>
+    <expand macro="requirements" />
+    <version_command><![CDATA[
+Rscript '$__tool_directory__/scripts/clusterinspect.R' @GET_VERSION@
+    ]]></version_command>
+    <command detect_errors="exit_code"><![CDATA[
+#if str($tool.mode) == "cluster"
+#set bin = 'cluster.R'
```
Comment (bgruening) [inline tools/raceid3/raceid3_main.xml]: here you can use indentation to make it more readable

File: tools/raceid3/raceid3_main.xml [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+#else if str($tool.mode) == "inspectclusters"
+#set bin = 'clusterinspect.R'
+#else if str($tool.mode) == "trajectory"
+#set bin = 'pseudotemporal.R'
+#else if str($tool.mode) == "inspecttrajectories"
+#set bin = 'trajectoryinspect.R'
+#end if
+
+Rscript $__tool_directory__/scripts/$bin '$userconf'
```
Comment (bgruening) [inline tools/raceid3/raceid3_main.xml]: single quotes

File: tools/raceid3/.shed.yml (hunk 2 above)
Comment (mtekman) [reply tools/raceid3/.shed.yml]: wouldn't all other single cell tools in the works (seurat, monocle) be grouped under the same suite?

File: tools/raceid3/macros.xml (hunk 3 above)
Comment (mtekman) [reply tools/raceid3/macros.xml]: I had some issues and so reverted back to not using this macro. But I think I know what the problem was, so I cab try incorporating this again

File: tools/raceid3/.shed.yml (hunk 2 above)
Comment (bgruening) [reply tools/raceid3/.shed.yml]: No. Only the tools in the reaceid folder. Indeed you only have one folder here, so the suite is pointless now. seurat and co will get an own folder ... we could at some point create a suite of all other tools .... but I would rather go for a tool.yaml file which servers the same purpose in the end.

File: tools/raceid3/macros_inspectclusters.xml (hunk 5 above)
Comment (mtekman) [reply tools/raceid3/macros_inspectclusters.xml]: I find that it is helpful to know what the defaults are for debugging purposes, so would prefer to keep them. But I can remove if it wastes too much space

File: tools/raceid/raceid_main.xml [hunk 12]
```diff
@@ -1,403 +1,202 @@
-<tool id="raceid_main" name="RaceID" version="@VERSION@.0">
-    <description>Race ID pipeline for single-cell RNA analysis</description>
+<tool id="raceid" name="RaceID" version="@VERSION_RACEID@.@VERSION_PACKAGE@.1" >
```
Comment (bgruening) [inline tools/raceid/raceid_main.xml]: can we keep this as "raceid_main" so that the upgrade path is not broken?

**Comment by mtekman** [comment]
@bgruening Yes, forgot to close this. It is superceded by #2170

**Pull request closed without merging**